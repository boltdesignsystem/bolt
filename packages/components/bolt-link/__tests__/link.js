import {
  render,
  renderString,
  stopServer,
  html,
} from '../../../testing/testing-helpers';
import schema from '../link.schema';
const { display, valign } = schema.properties;

const timeout = 90000;

describe('link', () => {
  let page;

  afterAll(async () => {
    await stopServer();
    await page.close();
  });

  beforeEach(async () => {
    await page.evaluate(() => {
      document.body.innerHTML = '';
    });
  }, timeout);

  beforeAll(async () => {
    page = await global.__BROWSER__.newPage();
    await page.goto('http://127.0.0.1:4444/', {
      timeout: 0,
    });
  }, timeout);

  test('basic link', async () => {
    const results = await render('@bolt-components-link/link.twig', {
      text: 'Hello World',
      url: 'https://pega.com',
    });
    expect(results.ok).toBe(true);
    expect(results.html).toMatchSnapshot();
  });

  display.enum.forEach(async option => {
    test(`link display: ${option}`, async () => {
      const results = await render('@bolt-components-link/link.twig', {
        text: 'Hello World',
        url: 'https://pega.com',
        display: option,
      });
      expect(results.ok).toBe(true);
      expect(results.html).toMatchSnapshot();
    });
  });

  valign.enum.forEach(async option => {
    test(`link valign: ${option}`, async () => {
      const results = await render('@bolt-components-link/link.twig', {
        text: 'Hello World',
        url: 'https://pega.com',
        valign: option,
      });
      expect(results.ok).toBe(true);
      expect(results.html).toMatchSnapshot();
    });
  });

  test('Link with outer classes via Drupal Attributes', async () => {
    const results = await render('@bolt-components-link/link.twig', {
      text: 'Link with outer classes',
      url: 'https://pega.com',
      attributes: {
        class: ['u-bolt-padding-medium'],
      },
    });
    expect(results.ok).toBe(true);
    expect(results.html).toMatchSnapshot();
  });

  test('Link with inner classes via Drupal Attributes', async () => {
    const results = await render('@bolt-components-link/link.twig', {
      text: 'Link with inner classes',
      url: 'https://pega.com',
      attributes: {
        class: ['is-active'],
      },
    });
    expect(results.ok).toBe(true);
    expect(results.html).toMatchSnapshot();
  });

  test('Link with outer JS-class via Drupal Attributes', async () => {
    const results = await render('@bolt-components-link/link.twig', {
      text: 'Link with outer JS-prefixed class',
      url: 'https://pega.com',
      attributes: {
        class: ['js-click-me'],
      },
    });
    expect(results.ok).toBe(true);
    expect(results.html).toMatchSnapshot();
  });

  test('Link with c-bolt- class is thrown out', async () => {
    const results = await render('@bolt-components-link/link.twig', {
      text: 'Link with outer JS-prefixed class',
      url: 'https://pega.com',
      attributes: {
        class: ['c-bolt-link--secondary'],
      },
    });
    expect(results.ok).toBe(true);
    expect(results.html).toMatchSnapshot();
  });

  test('Link with an on_click param renders properly', async () => {
    const results = await render('@bolt-components-link/link.twig', {
      text: 'Link with on_click via param',
      url: 'https://pega.com',
      on_click: 'on-click-test',
    });
    expect(results.ok).toBe(true);
    expect(results.html).toMatchSnapshot();
  });

  test('Link with an on_click attributes renders properly', async () => {
    const results = await render('@bolt-components-link/link.twig', {
      text: 'Link w/ on_click via attributes',
      url: 'https://pega.com',
      attributes: {
        'on-click': 'on-click-test',
      },
    });
    expect(results.ok).toBe(true);
    expect(results.html).toMatchSnapshot();
  });

  test('Default <bolt-link> w/o Shadow DOM renders', async function() {
    const renderedLinkHTML = await page.evaluate(async () => {
      document.body.insertAdjacentHTML(
        'beforeend',
        '<div><bolt-link url="http://pega.com" no-shadow>This is a link</bolt-link></div>',
      );
      const link = document.querySelector('bolt-link');
      await customElements.whenDefined('bolt-link');
      await link.updateComplete;
      return link.parentNode.outerHTML;
    });
    expect(renderedLinkHTML).toMatchSnapshot();

    const renderedHTML = await html(renderedLinkHTML);

    expect(
      renderedHTML
        .querySelector('.c-bolt-link')
        .classList.contains('c-bolt-link--display-inline'),
    ).toBe(true);

    expect(renderedHTML).toMatchSnapshot();
  });

  test('Default <bolt-link> with Shadow DOM renders', async function() {
    const defaultLinkShadowRoot = await page.evaluate(async () => {
      document.body.insertAdjacentHTML(
        'beforeend',
        '<bolt-link url="http://pega.com">Link Test -- Shadow Root HTML</bolt-link>',
      );
      const link = document.querySelector('bolt-link');
      await customElements.whenDefined('bolt-link');
      await link.updateComplete;
      return link.renderRoot.innerHTML;
    });
    expect(defaultLinkShadowRoot).toMatchSnapshot();

    const defaultLinkOuter = await page.evaluate(async () => {
      const link = document.createElement('bolt-link');
      link.setAttribute('url', 'http://pega.com');
      link.textContent = 'Link Test -- Outer HTML';
      document.body.appendChild(link);
      await customElements.whenDefined('bolt-link');
      await link.updateComplete;
      return link.outerHTML;
    });
    expect(defaultLinkOuter).toMatchSnapshot();

    const renderedHTML = await html(defaultLinkOuter);
    expect(renderedHTML.textContent).toEqual('Link Test -- Outer HTML');
  });

  test('Default <bolt-link> w/o Shadow DOM renders and without url prop', async function() {
    const renderedLinkHTML = await page.evaluate(async () => {
      document.body.insertAdjacentHTML(
        'beforeend',
        '<div><bolt-link no-shadow>This is a link without url prop' +
          '</bolt-link></div>',
      );
      const link = document.querySelector('bolt-link');
      await customElements.whenDefined('bolt-link');
      await link.updateComplete;
      return link.parentNode.outerHTML;
    });
    expect(renderedLinkHTML).toMatchSnapshot();

    const renderedHTML = await html(renderedLinkHTML);

    expect(renderedHTML.hasAttribute('href')).toBe(false);
    expect(
      renderedHTML
        .querySelector('.c-bolt-link')
        .classList.contains('c-bolt-link--display-inline'),
    ).toBe(true);

    expect(renderedHTML).toMatchSnapshot();
  });

  test('Default <bolt-link> with Shadow DOM renders and without url prop', async function() {
    const defaultLinkShadowRoot = await page.evaluate(async () => {
      document.body.insertAdjacentHTML(
        'beforeend',
        '<bolt-link>Link Test without url prop -- Shadow Root HTML</bolt-link>',
      );
      const link = document.querySelector('bolt-link');
      await customElements.whenDefined('bolt-link');
      await link.updateComplete;
      return link.renderRoot.innerHTML;
    });
    expect(defaultLinkShadowRoot).toMatchSnapshot();

    const defaultLinkOuter = await page.evaluate(async () => {
      const link = document.createElement('bolt-link');
      link.textContent = 'Link Test without url attr -- Outer HTML';
      document.body.appendChild(link);
      await customElements.whenDefined('bolt-link');
      await link.updateComplete;
      return link.outerHTML;
    });
    expect(defaultLinkOuter).toMatchSnapshot();

    const renderedHTML = await html(defaultLinkOuter);
    expect(renderedHTML.hasAttribute('href')).toBe(false);
    expect(renderedHTML.textContent).toEqual(
      'Link Test without url attr -- Outer HTML',
    );
  });

  test('Default <bolt-link> with Shadow DOM renders with no extra whitespace', async function() {
    const defaultLinkOuter = await page.evaluate(async () => {
      // Include huge inline font-size style to increase visibility of any unexpected whitespace
      document.body.insertAdjacentHTML(
        'beforeend',
        '<div style="font-size: 300%;">(<bolt-link url="http://pega.com">Link Test -- No extra whitespace</bolt-link>)</div>',
      );
      const link = document.querySelector('bolt-link').parentNode;
      await customElements.whenDefined('bolt-link');
      await link.updateComplete;
      return link.outerHTML;
    });
    expect(defaultLinkOuter).toMatchSnapshot();

    const renderedHTML = await html(defaultLinkOuter);

    expect(renderedHTML.querySelector('bolt-link').textContent).toEqual(
      'Link Test -- No extra whitespace',
    );

    expect(renderedHTML).toMatchSnapshot();
  });

  test('<bolt-link> does not have duplicate IDs', async () => {
    // The reported bug was only a problem when rendering the initial link with
    // twig, so start by rendering the link with twig.
    const template = await renderString(`
      {% include "@bolt-components-link/link.twig" with {
        text: "Test link -- Has id attribute",
        url: "https://pega.com",
        attributes: {
          id: 'my-link'
         }
        } only %}
    `);

    // Next, convert to a javascript node and disable shadow dom so we can evaluate it with js.
    const renderedLinkHTML = await page.evaluate(async html => {
      const div = document.createElement('div');
      document.body.insertAdjacentHTML('beforeend', html);
      const link = document.querySelector('bolt-link');
      await customElements.whenDefined('bolt-link');
      await link.updateComplete;
      return link.outerHTML;
    }, template.html);

    const renderedHTML = await html('<div></div>');
    renderedHTML.innerHTML = renderedLinkHTML;

    // Loop though all elements with any ID to see if there are duplicates.
    // todo: `querySelectorAll('#my-id')` does not return multiple elements when there are dupes. It does in Chrome. See if upgrading basichtml from "0.22.1" to "1.1.1" adds that feature.
    const elemsWithAnyId = renderedHTML.querySelectorAll('[id]');
    let allIds = [];
    let dupIds = [];
    elemsWithAnyId.forEach(el => {
      if (el.id) {
        if (allIds.includes(el.id)) {
          dupIds.push(el.id);
        } else {
          allIds.push(el.id);
        }
      }
    });

    expect(dupIds.length).toEqual(0);
  });
});
