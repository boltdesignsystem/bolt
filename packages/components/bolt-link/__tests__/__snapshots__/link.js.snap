// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`link Default <bolt-link> w/o Shadow DOM renders 1`] = `
<div>
  <bolt-link url="http://pega.com"
             no-shadow
  >
    <a href="http://pega.com"
       target="_self"
       class="c-bolt-link c-bolt-link--display-inline c-bolt-link--valign-center"
    >
      <slot name="before">
      </slot>
      <span class="c-bolt-link__text">
        This is a link
      </span>
      <slot name="after">
      </slot>
    </a>
  </bolt-link>
</div>
`;

exports[`link Default <bolt-link> w/o Shadow DOM renders 2`] = `
<div
  style=""
>
  <bolt-link
    no-shadow=""
    style=""
    url="http://pega.com"
  >
    <!---->
    <!---->
  </bolt-link>
  <a
    class="c-bolt-link c-bolt-link--display-inline c-bolt-link--valign-center"
    href="http://pega.com"
    style=""
    target="_self"
  >
    <!---->
    <!---->
    <slot
      name="before"
      style=""
    />
    <!---->
    <span
      class="c-bolt-link__text"
      style=""
    >
      <!---->
      <!---->
      This is a link
      <!---->
      <!---->
    </span>
    <!---->
    <slot
      name="after"
      style=""
    />
    <!---->
    <!---->
  </a>
  <!---->
  <!---->
</div>
`;

exports[`link Default <bolt-link> w/o Shadow DOM renders and without url prop 1`] = `
<div>
  <bolt-link no-shadow>
    <button class="c-bolt-link c-bolt-link--display-inline c-bolt-link--valign-center">
      <slot name="before">
      </slot>
      <span class="c-bolt-link__text">
        This is a link without url prop
      </span>
      <slot name="after">
      </slot>
    </button>
  </bolt-link>
</div>
`;

exports[`link Default <bolt-link> w/o Shadow DOM renders and without url prop 2`] = `
<div
  style=""
>
  <bolt-link
    no-shadow=""
    style=""
  >
    <!---->
    <!---->
  </bolt-link>
  <button
    class="c-bolt-link c-bolt-link--display-inline c-bolt-link--valign-center"
    style=""
  >
    <!---->
    <!---->
    <slot
      name="before"
      style=""
    />
    <!---->
    <span
      class="c-bolt-link__text"
      style=""
    >
      <!---->
      <!---->
      This is a link without url prop
      <!---->
      <!---->
    </span>
    <!---->
    <slot
      name="after"
      style=""
    />
    <!---->
    <!---->
  </button>
  <!---->
  <!---->
</div>
`;

exports[`link Default <bolt-link> with Shadow DOM renders 1`] = `
<a href="http://pega.com"
   target="_self"
   class="c-bolt-link c-bolt-link--display-inline c-bolt-link--valign-center"
>
  <slot name="before">
  </slot>
  <span class="c-bolt-link__text">
    <slot>
    </slot>
  </span>
  <slot name="after">
  </slot>
</a>
`;

exports[`link Default <bolt-link> with Shadow DOM renders 2`] = `
<bolt-link url="http://pega.com">
  Link Test -- Outer HTML
</bolt-link>
`;

exports[`link Default <bolt-link> with Shadow DOM renders and without url prop 1`] = `
<button class="c-bolt-link c-bolt-link--display-inline c-bolt-link--valign-center">
  <slot name="before">
  </slot>
  <span class="c-bolt-link__text">
    <slot>
    </slot>
  </span>
  <slot name="after">
  </slot>
</button>
`;

exports[`link Default <bolt-link> with Shadow DOM renders and without url prop 2`] = `
<bolt-link>
  Link Test without url attr -- Outer HTML
</bolt-link>
`;

exports[`link Default <bolt-link> with Shadow DOM renders with no extra whitespace 1`] = `
<div style="font-size: 300%;">
  (
  <bolt-link url="http://pega.com">
    Link Test -- No extra whitespace
  </bolt-link>
  )
</div>
`;

exports[`link Default <bolt-link> with Shadow DOM renders with no extra whitespace 2`] = `
<div
  style="font-size:300%;"
>
  (
  <bolt-link
    style=""
    url="http://pega.com"
  >
    Link Test -- No extra whitespace
  </bolt-link>
  )
</div>
`;

exports[`link Link with an on_click attributes renders properly 1`] = `
<bolt-link on-click="on-click-test"
           url="https://pega.com"
>
  <a href="https://pega.com"
     target="_self"
     class="c-bolt-link c-bolt-link--display-inline c-bolt-link--valign-center"
  >
    <ssr-keep for="bolt-link"
              class="c-bolt-link__text"
    >
      Link w/ on_click via attributes
    </ssr-keep>
  </a>
</bolt-link>
`;

exports[`link Link with an on_click param renders properly 1`] = `
<bolt-link url="https://pega.com"
           on-click="on-click-test"
>
  <a href="https://pega.com"
     target="_self"
     class="c-bolt-link c-bolt-link--display-inline c-bolt-link--valign-center"
  >
    <ssr-keep for="bolt-link"
              class="c-bolt-link__text"
    >
      Link with on_click via param
    </ssr-keep>
  </a>
</bolt-link>
`;

exports[`link Link with c-bolt- class is thrown out 1`] = `
<bolt-link url="https://pega.com">
  <a href="https://pega.com"
     target="_self"
     class="c-bolt-link c-bolt-link--display-inline c-bolt-link--valign-center"
  >
    <ssr-keep for="bolt-link"
              class="c-bolt-link__text"
    >
      Link with outer JS-prefixed class
    </ssr-keep>
  </a>
</bolt-link>
`;

exports[`link Link with inner classes via Drupal Attributes 1`] = `
<bolt-link url="https://pega.com">
  <a href="https://pega.com"
     target="_self"
     class="c-bolt-link c-bolt-link--display-inline c-bolt-link--valign-center is-active"
  >
    <ssr-keep for="bolt-link"
              class="c-bolt-link__text"
    >
      Link with inner classes
    </ssr-keep>
  </a>
</bolt-link>
`;

exports[`link Link with outer JS-class via Drupal Attributes 1`] = `
<bolt-link url="https://pega.com"
           class="js-click-me"
>
  <a href="https://pega.com"
     target="_self"
     class="c-bolt-link c-bolt-link--display-inline c-bolt-link--valign-center"
  >
    <ssr-keep for="bolt-link"
              class="c-bolt-link__text"
    >
      Link with outer JS-prefixed class
    </ssr-keep>
  </a>
</bolt-link>
`;

exports[`link Link with outer classes via Drupal Attributes 1`] = `
<bolt-link url="https://pega.com"
           class="u-bolt-padding-medium"
>
  <a href="https://pega.com"
     target="_self"
     class="c-bolt-link c-bolt-link--display-inline c-bolt-link--valign-center"
  >
    <ssr-keep for="bolt-link"
              class="c-bolt-link__text"
    >
      Link with outer classes
    </ssr-keep>
  </a>
</bolt-link>
`;

exports[`link basic link 1`] = `
<bolt-link url="https://pega.com">
  <a href="https://pega.com"
     target="_self"
     class="c-bolt-link c-bolt-link--display-inline c-bolt-link--valign-center"
  >
    <ssr-keep for="bolt-link"
              class="c-bolt-link__text"
    >
      Hello World
    </ssr-keep>
  </a>
</bolt-link>
`;

exports[`link link display: block 1`] = `
<bolt-link url="https://pega.com"
           display="block"
>
  <a href="https://pega.com"
     target="_self"
     class="c-bolt-link c-bolt-link--display-block c-bolt-link--valign-center"
  >
    <ssr-keep for="bolt-link"
              class="c-bolt-link__text"
    >
      Hello World
    </ssr-keep>
  </a>
</bolt-link>
`;

exports[`link link display: flex 1`] = `
<bolt-link url="https://pega.com"
           display="flex"
>
  <a href="https://pega.com"
     target="_self"
     class="c-bolt-link c-bolt-link--display-flex c-bolt-link--valign-center"
  >
    <ssr-keep for="bolt-link"
              class="c-bolt-link__text"
    >
      Hello World
    </ssr-keep>
  </a>
</bolt-link>
`;

exports[`link link display: inline 1`] = `
<bolt-link url="https://pega.com"
           display="inline"
>
  <a href="https://pega.com"
     target="_self"
     class="c-bolt-link c-bolt-link--display-inline c-bolt-link--valign-center"
  >
    <ssr-keep for="bolt-link"
              class="c-bolt-link__text"
    >
      Hello World
    </ssr-keep>
  </a>
</bolt-link>
`;

exports[`link link valign: center 1`] = `
<bolt-link url="https://pega.com"
           valign="center"
>
  <a href="https://pega.com"
     target="_self"
     class="c-bolt-link c-bolt-link--display-inline c-bolt-link--valign-center"
  >
    <ssr-keep for="bolt-link"
              class="c-bolt-link__text"
    >
      Hello World
    </ssr-keep>
  </a>
</bolt-link>
`;

exports[`link link valign: start 1`] = `
<bolt-link url="https://pega.com"
           valign="start"
>
  <a href="https://pega.com"
     target="_self"
     class="c-bolt-link c-bolt-link--display-inline c-bolt-link--valign-start"
  >
    <ssr-keep for="bolt-link"
              class="c-bolt-link__text"
    >
      Hello World
    </ssr-keep>
  </a>
</bolt-link>
`;
